src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.State.cs

```diff
@@ -0,0 +1,296 @@
+                if (root.DescendantNodes().OfType<TNamespaceDeclarationSyntax>().Count() > 1)
+                {
+                    return null;
+                }
+
+                // Bail if cursor is not on namespace declaration.
+                // TODO: Note that this also means we will ignore document with only global namespace.
+                var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+                if (!syntaxFacts.IsOnNamespaceDeclaration(root, textSpan.Start))
+                {
+                    return null;
+                }
+
+                var rootNamespace = await GetRootNamespaceAsync(document.Project, cancellationToken).ConfigureAwait(false);
+                if (rootNamespace == null)
+                {
+                    return null;
+                }
+
+                var namespaceDeclaration = root.FindToken(textSpan.Start).GetAncestor<TNamespaceDeclarationSyntax>();
+                Debug.Assert(namespaceDeclaration != null);
+
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false); 
+                var namespaceSymbol = semanticModel.GetDeclaredSymbol(namespaceDeclaration, cancellationToken) as INamespaceSymbol;
+                if (namespaceSymbol == null)
+                {
+                    return null;
+                }
+
+                var declaredNamespace = namespaceSymbol.ToDisplayString(QualifiedNameOnlyFormat);
+
+                // Namespace can't be renamed if we can't construct a valid qualified identifier from folder names.
+                // In this case, we might still be able to provide refactoring to move file to new location.
+                var targetNamespace = TryBuildNamespaceFromFolders(document.Folders, syntaxFacts, out var qualifiedIdentifierFromFolders) 
+                                        ? ConcatNamespace(rootNamespace, qualifiedIdentifierFromFolders)
+                                        : null;
+
+                // No action required if namespace already matches folders.
+                if (string.Equals(targetNamespace, declaredNamespace, System.StringComparison.Ordinal))
+                {
+                    return null;
+                }
+
+                // Only provide "move file" action if root namespace is a proper prefix of declared namespace.
+                IsProperPrefix(declaredNamespace, rootNamespace, out var qualifiedIdentifierFromDeclaration);
+                                                                           
+                // Only provide "change namespace" action if it wouldn't cause conflicts in target namespace.
+                if (targetNamespace != null && 
+                    (await WillCauseConflictInTargetNamespace(document.Project, targetNamespace, service.GetDeclaredSymbols(namespaceDeclaration, semanticModel, cancellationToken), cancellationToken).ConfigureAwait(false)))
```

Comment:
too long to fit on a single line (And in an if-condition to boot).  please break out